Risky orders browser

reinaris opened the pull request:
I created a listing for risky orders.
This listing should be useful for customer service people who have to handle risky orders on a daily base.
## Risky order listing

![screen_shot_2015-02-04_at_17_35_01](https://cloud.githubusercontent.com/assets/106335/6045738/7ac28e8a-ac94-11e4-8981-169924df5ac0.png)

The listing generates a quick view of the order on the right side of the screen.
You can approve or cancel the order.
After the action the order disappears from the list and you will automatically jump to the next order.

The idea is that you can't start with the first and just work your way down to the last by simply clicking accept or cancel.

Need to do:
1) Write specs

I like to hear your opinion about this screen!

reinaris committed:
Improved layout for pagination and per-page listing options

reinaris committed:
Added approve method to the order API

reinaris committed:
Css for risky orders listing

reinaris committed:
Added risky orders to the mainmenu

reinaris committed:
Added the risky listing templates & javascript

reinaris committed:
Improved the risky orders layout

reinaris committed:
Get rid of the determine_layout method

reinaris committed:
Improved risky orders javascript

reinaris committed:
Improvements of risky orders javascript

reinaris committed:
Improved risky orders information

reinaris committed:
Improved risky orders javascript

reinaris committed:
Improved risky orders javascript

reinaris committed:
Replaced where query with scope

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://reinaris-risky-orders-browser.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://reinaris-risky-orders-browser.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

jhawthorn commented on api/app/controllers/spree/api/orders_controller.rb:
```diff
@@ -14,6 +14,12 @@ class OrdersController < Spree::Api::BaseController
       end
 
+      def approve
+        authorize! :update, @order, params[:token]
```
:bomb: This isn't a sufficient check. All users have the ability to update their own order. Only admins should be able to approve an order.

reinaris replied:
@jhawthorn thanks for the response!
I'm not sure, but the way to do this is `authorize! :admin, @order, params[:token]` right?

simmsy replied:
@reinaris This can be achieved with a new custom action so `authorize! :approve, @order`

By default admin has the ability to manage all objects and therefore would be able to approve this but user's can only update so will not.

JDutil replied:
:+1: on @simmsy suggestion for a new action @reinaris

reinaris committed:
Improved some code syntax

reinaris committed:
Improved risky orders code quality

reinaris committed:
Authorize users before approving an order trough API

reinaris committed:
Improved risky orders javascript

Handle situation when all risky orders are processed

reinaris committed:
Removed the action from per-page-form

So it will always submit to the same page.
For example /orders/risky would have been submitted to /orders (index)
which is not right. Now it submits to the current page /orders/risky.

reinaris committed:
Improved risky javascript code

peterberkenbosch commented:
So the risky orders would also list the not-risky orders? Confused with the screenshot. Could this not be just a small addition on the standard order list page? Filter risky orders.

reinaris commented:
@peterberkenbosch it only shows risky orders.
Sorry for confusing you, I had a messed up dataset :) But imagine all orders are risky.

It could be an feature which shows up when you filters the order index on risky orders.
This quickview feature on the right side is also very handy for non risky orders.

reinaris committed:
improved risky orders javascript

More clear way to say ‘not true’

JDutil commented:
I mentioned this to @simmsy offline that I'm thinking we should just have the order preview piece of this directly on the order index action rather than creating a new action.  I think the order preview piece of it is useful for more than just risky orders like customer service etc...  The preview button could expand/collapse the order preview section for approving/canceling risky orders, and users can simply use the advanced filter form to filter only risky orders.

I'm not sure if that would make the page too cluttered or not, but I think it makes sense.

reinaris commented:
@JDutil this preview piece is very handy for the orders index, so yes!
On the orders index you can filter on risky, if the preview piece is there we achieve the same as the risky order browser is doing right now.

So If I have you vote I will create another PR for the order preview piece.
Then we could close this one.
Let me know if thats the way to go.

JDutil commented:
@reinaris that is what I'd prefer to do, and what I believe most people would like. You may want to ping @simmsy about it since he said he'd like to think on it, but I'm thinking that's the way to go. It also seemed like the direction you guys were wanting to take in the admin working group when you were discussing ways to improve the product management.

JDutil commented:
It also reduces code duplication FTW ;)

reinaris commented:
Agree! Reduces a lot of code, so best solution indeed.
Will be hacking on this later..

simmsy commented:
Guys it works for me so long as the behaviour that removes risky orders from the list view when approved still remains. I believe that there may be some uncertainty to overcome on elegantly manipulating a filtered list view that is not explicitly pre defined as being risky, and therefore if this can be overcome I fully support this. Also the preview may need to become state aware and therefore also easily customisable in each state if desired, again an uncertainty that was not probably considered in the initial implementation.

Sent from my iPhone

On 18 Feb 2015, at 19:04, Rein Aris <notifications@github.com<mailto:notifications@github.com>> wrote:


Agree! Reduces a lot of code, so best solution indeed.
Will be hacking on this later..

—
Reply to this email directly or view it on GitHub<https://github.com/spree/spree/pull/6006#issuecomment-74925553>.

bbonislawski commented:
This is stale and I'm closing this for now. If anyone thinks that it should be reopen and want to take care of this PR, please mention me
